### PR TITLE
Link: add TapArea-like feedback

### PR DESCRIPTION
**Pull request by jackhsu978**
### Minor changes
- Add `<TapArea>`-like touch feedback to `<Link>`

### Revisions
- minor fix for a test name in `TapArea`

## Details

Here is a screenshot of the API documentation:

Screenshot 2020-06-19 18 09 58

It supports some of props that `<TapArea>` supports, but not all since not every option is relevant to `<TapLink>`.

Here is a screenshot of examples in the documentation:

Screenshot 2020-06-19 18 10 58

Note that the "Prevent default on tap" switch is turn on by default because it would be easier to test the TapArea-like touch feedback without actually opening the link. One can turn it off and test the link navigation.

## TODO

- [x] Accessibility checkup
- [x] Update documentation
- [x] Add/update Tests
- [ ] Pinterest Designer (for design changes)
- [x] [Format PR title](https://github.com/pinterest/gestalt/#releasing): `ComponentName: Description`.

### CODE REVIEW

File: packages/gestalt/src/TapLink.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+          event.preventDefault();
+          onTap({ event });
+        }
+      }}
+      onTouchStart={handleTouchStart}
+      onTouchMove={handleTouchMove}
+      onTouchCancel={handleTouchCancel}
+      onTouchEnd={handleTouchEnd}
+      rel={[
+        ...(target === 'blank' ? ['noopener', 'noreferrer'] : []),
```
Comment (christianvuerings) [inline packages/gestalt/src/TapLink.js:88]: @jackhsu978 I like this a lot - we should follow the same logic in `<Link />` (we should do that in a follow up PR)

File: docs/src/TapLink.doc.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  return (
+    <Box>
+      <Box display="flex" alignItems="center" padding={1} marginStart={-2} marginEnd={-2}>
+        <Box paddingX={2}>
+          <Label htmlFor="preventDefault">
+            <Text>Prevent default on tap</Text>
+          </Label>
+        </Box>
+        <Switch
+          preventDefault
```
Comment (christianvuerings) [inline docs/src/TapLink.doc.js:85]: @jackhsu978 should this have been `id="preventDefault"` to link the label to the switch?

File: docs/src/TapLink.doc.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        >
+          <Text>TapLink</Text>
+        </TapLink>
+      </Box>
+      <Box padding={1}>
+        <TapLink
+          href="https://pinterest.com"
+          onTap={onTap}
+          rounding="pill"
+          tapStyle="compress"
```
Comment (christianvuerings) [inline docs/src/TapLink.doc.js:105]: @jackhsu978 could we make this example also `inline` - `compress` with full width feels pretty odd

File: docs/src/TapLink.doc.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+          '({ event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticKeyboardEvent<HTMLAnchorElement> }) => void',
+        href: 'basicExample',
+      },
+      {
+        name: 'rel',
+        type: `"none" | "nofollow"`,
+      },
+      {
+        name: 'rounding',
+        type: `"pill" | "circle" | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8`,
```
Comment (christianvuerings) [inline docs/src/TapLink.doc.js:51]: @jackhsu978 could we mention that the default value is `0`?

File: docs/src/TapLink.doc.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        href: 'basicExample',
+      },
+      {
+        name: 'target',
+        type: `"null" | "self" | "blank"`,
+        defaultValue: 'null',
+      },
+    ]}
```
Comment (christianvuerings) [inline docs/src/TapLink.doc.js:59]: @jackhsu978 could we also add `tapStyle` to the docs?

File: docs/src/TapLink.doc.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      },
+      {
+        name: 'onTap',
+        type:
+          '({ event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticKeyboardEvent<HTMLAnchorElement> }) => void',
+        href: 'basicExample',
+      },
+      {
+        name: 'rel',
+        type: `"none" | "nofollow"`,
```
Comment (christianvuerings) [inline docs/src/TapLink.doc.js:47]: @jackhsu978 could we mention that the default is `none`?

File: docs/src/TapLink.doc.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  const [preventDefault, setPreventDefault] = React.useState(true);
+  const onTap = ({ event }) => {
+    if (preventDefault) {
+      event.preventDefault();
+    }
+  };
+
+  return (
+    <Box>
+      <Box display="flex" alignItems="center" padding={1} marginStart={-2} marginEnd={-2}>
```
Comment (christianvuerings) [inline docs/src/TapLink.doc.js:78]: @jackhsu978 seems like a good use case for `Row`

```jsx
<Row gap={2} padding={1}>
  <Label htmlFor="preventDefault">
    <Text>Prevent default on tap</Text>
  </Label>
  <Switch
    id="preventDefault"
    onChange={() => setPreventDefault(!preventDefault)}
    switched={preventDefault}
  />
</Row>
```

**Comment by ashleyseto** [comment]
@ajmoo will be reviewing this from PDS side and all other tap related PRs

**Comment by ajmoo** [comment]
For context… is the purpose of this component to support mWeb? Would love to understand a bit more around why we have both Link and TapLink components. Why not add this functionality to the Link component? Forgive me if these are dumb questions.

**Comment by jackhsu978** [comment]
@christianvuerings @bishwei do you think it makes sense to add the features to Link rather than having a new component?

File: docs/src/TapLink.doc.js (hunk 3 above)
Comment (jackhsu978) [reply docs/src/TapLink.doc.js:121]: Sure, updated. See screenshot below:

Screenshot 2020-06-19 13 47 48

File: packages/gestalt/src/TapLink.js (hunk 1 above)
Comment (jackhsu978) [reply packages/gestalt/src/TapLink.js:88]: sure

File: docs/src/TapLink.doc.js (hunk 2 above)
Comment (jackhsu978) [reply docs/src/TapLink.doc.js]: good catch! done!

**Comment by jackhsu978** [comment]
@ajmoo this is a great question! I talked to @christianvuerings offline and we think that we should just merge these to `<Link>` component rather than creating a new component. We just need to make sure that the `<Link>` today works the same way.

PS: Please feel free to ask any questions here. It can also help others who have the same questions. You can always DM me if you feel more comfortable doing that.

**Comment by jackhsu978** [comment]
@ajmoo @ashleyseto @christianvuerings Thank you for all the great feedback. I've changed the PR to add additional feature to `<Link>` rather than creating a new component. Please take another look when you get a chance ([test link](https://deploy-preview-933--gestalt.netlify.app/#/Link)). Thank you!

File: docs/src/Link.doc.js [hunk 8]
```diff
@@ -111,4 +138,68 @@ card(
...
+  const onClick = ({ event }) => {
+    if (preventDefault) {
+      event.preventDefault();
+    }
+  };
+
+  return (
+    <Box>
+      <Row gap={2} padding={1}>
+        <Box paddingX={2}>
```
Comment (christianvuerings) [inline docs/src/Link.doc.js:157]: @jackhsu978 now that we use `<Row gap={2} />`, we no longer need this `<Box paddingX={2}>`.

I would change it to the following:
```
      <Row gap={2} padding={2}>
        <Label htmlFor="preventDefault">
          <Text>Prevent default on tap</Text>
        </Label>
        <Switch
          id="preventDefault"
          onChange={() => setPreventDefault(!preventDefault)}
          switched={preventDefault}
        />
      </Row>
      <Divider />
```

To fix alignment + make it obvious that the toggle changes the examples below.

Comment (jackhsu978) [reply docs/src/Link.doc.js]: done. thank you for suggestions!

**Pull request merged**